### PR TITLE
Refactor Cthulhu to allow custom callinfo for external packages

### PR DESCRIPTION
**Pull request by Keno**
This is intended to allow Diffactor to provide a Cthulhu interfaces
for navigating between different differentiation levels. The core
interface is the introduction of `Cthulhu.AbstractCursor`, which
downstream packages can extend to provide additional information
beyond an mi to identify the frame that's currently being displayed.
These interfaces will probably need to evolve a bit, but I think this
is a reasonable start.

Also strengthen the tests a bit to make failures more obvious and
easier to debug.

### CODE REVIEW

**Review by vchuravy — APPROVED** [review]
Maybe we eventually need to factor this out to a Interface package?

Then we could move the GPU integration to GPUCompiler

**Comment by Keno** [comment]
> Maybe we eventually need to factor this out to a Interface package?

Yes, I think we'll want to go this way, but we should probably refine the interface first.

File: src/interface.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    mi::MethodInstance
+end
+
+function update_mi(curs::CthulhuCursor, mi::MethodInstance)
+    CthulhuCursor(mi)
+end
+
+# This method is optional, but should be implemented if there is
+# a sensible default cursor for a MethodInstance
+AbstractCursor(interp::CthulhuInterpreter, mi::MethodInstance) =
```
Comment (aviatesk) [inline src/interface.jl:12]:
```suggestion
AbstractCursor(interp::AbstractInterpreter, mi::MethodInstance) =
```

Comment (aviatesk) [reply src/interface.jl:12]: Should make it optional.

File: src/interface.jl [hunk 2]
```diff
@@ -0,0 +1,36 @@
+abstract type AbstractCursor; end
+struct CthulhuCursor <: AbstractCursor
+    mi::MethodInstance
+end
+
+function update_mi(curs::CthulhuCursor, mi::MethodInstance)
+    CthulhuCursor(mi)
+end
```
Comment (aviatesk) [inline src/interface.jl:8]:
```suggestion
update_mi(curs::AbstractCursor, ::MethodInstance) = error("""
missing `$AbstractCursor` API:
`$(typeof(curs))` is required to implement the `$update_mi(::$(typeof(curs)), ::MethodInstance) -> $(typeof(curs))` interface.
""")
update_mi(curs::CthulhuCursor, mi::MethodInstance) = CthulhuCursor(mi)
```

File: src/interface.jl [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# This method is optional, but should be implemented if there is
+# a sensible default cursor for a MethodInstance
+AbstractCursor(interp::CthulhuInterpreter, mi::MethodInstance) =
+    CthulhuCursor(mi)
+
+function get_optimized_code(interp::CthulhuInterpreter,
+                                curs::CthulhuCursor)
+    return interp.opt[curs.mi].inferred
+end
+
```
Comment (aviatesk) [inline src/interface.jl:19]:
```suggestion
```
Never used.

**Comment by Keno** [comment]
I have some additional changes on top of this, but I think rather than ping-ponging this PR back and forth, I'm gonna merge this as is and then do my additional changes in a quick separate PR.

**Pull request merged**